File: pset6/pset6/credit/credit.py
# Get user input
card_num = str(input("Number: "))

# Initialize sum counter and get length
card_sum = 0
length = len(card_num)

# Loop through each digit
# j indexes the position from the right-hand side
# If on an even position, add the digits of twice the jth digit to the sum
# Otherwise, add the digit to the sum
for i in range(length):
    j = length - i - 1
    if not i % 2:
        card_sum += int(card_num[j])
    else:
        digit = 2 * int(card_num[j])
        if digit > 9:
            card_sum += (1 + digit % 10)
        else:
            card_sum += digit

# Check the sum for validity, and if valid, check the first one or two digits and the length
if card_sum % 10:
    print("INVALID")
elif length == 15 and (card_num[0:2] == "34" or card_num[0:2] == "37"):
    print("AMEX")
elif length == 16 and int(card_num[0:2]) > 50 and int(card_num[0:2]) <= 55:
    print("MASTERCARD")
elif length == 13 or length == 16 and card_num[0] == "4":
    print("VISA")
else:
    print("INVALID")

File: pset1/pset1/credit/credit.c
#include <stdio.h>
#include <cs50.h>

int main(void)
{
    long number = get_long("Please enter the card number: ");                                   //get long card number
    int length = 0;
    int sum = 0; 
    bool even = 0;
    int lastTwo;
    while (number > 0)                                                                          //loop through each digit
    {
        int digit = number % 10;                                                                //extract current ones digit
        number = number / 10;                                                                   //move everything to the right one place
        if (even)                                                                               //if place is even, multiply by 2
        {
            digit = 2 * digit;
        }
        if (digit > 9)                                                                          //if above 10, add 1 + X for 1X 
        {
            sum = sum + 1 + (digit % 10);
        }
        else
        {
            sum += digit;                                                                       //otherwise, just add X
        }
        length ++;                                                                              //keep track of length
        even = !even;                                                                           //next digit will be in opposite spot (even or odd)
        if (number < 100 && number >= 10)                                                       //keep track of last two digits for ID later
        {
            lastTwo = number;
        }
    }
    
    if (sum % 10)                                                                               //always invalid if it doesn't meet the sum criterion
    {
        printf("INVALID\n");
    }
    else if ((length == 15) && (lastTwo == 34 || lastTwo == 37))                                //check for AMEX
    {
        printf("AMEX\n");
    }
    else if (length == 16 && lastTwo > 50 && lastTwo <= 55)                                     //check for MasterCard
    {
        printf("MASTERCARD\n");
    }
    else if ((length == 13 || length == 16) && ((lastTwo / 10) == 4))                           //check for Visa
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");                                                                    //invalid otherwise
    }
}